contracts/test/UserRegistry.t.sol

```diff
@@ -0,0 +1,277 @@
+        uint256 tokenId = proxy.register(
+            "user1domain",
+            user2,
+            IRegistry(address(0)),
+            address(0),
+            ROLE_SET_SUBREGISTRY | ROLE_SET_RESOLVER,
+            uint64(block.timestamp + 365 days)
+        );
+        
+        // Verify registration was successful
+        assertEq(proxy.ownerOf(tokenId), user2, "Domain should be owned by user2");
+    }
+    
+    function test_Revert_unauthorized_registration() public {
+        // User1 tries to register a domain without ROLE_REGISTRAR
+        vm.prank(user1);
+        vm.expectRevert(); // Should revert because user1 doesn't have ROLE_REGISTRAR
+        proxy.register(
+            "unauthorizeddomain",
+            user1,
+            IRegistry(address(0)),
+            address(0),
+            ROLE_SET_SUBREGISTRY | ROLE_SET_RESOLVER,
+            uint64(block.timestamp + 365 days)
+        );
+    }
+    
+    function test_Revert_unauthorized_role_grant() public {
+        // User1 tries to grant roles without permission
+        vm.prank(user1);
+        vm.expectRevert(); // Should revert because user1 doesn't have ROLE_REGISTRAR_ADMIN
+        proxy.grantRootRoles(ROLE_REGISTRAR, user2);
+    }
+    
+    function testFuzz_domain_registration(string memory label, uint64 duration) public {
+        // Skip empty labels and ensure reasonable duration
+        vm.assume(bytes(label).length > 0);
+        duration = uint64(bound(duration, 1 days, 10 * 365 days));
+        
+        uint64 expires = uint64(block.timestamp) + duration;
+        
+        // Register a domain as admin
+        vm.prank(admin);
+        uint256 tokenId = proxy.register(
+            label,
+            user1,
+            IRegistry(address(0)),
+            address(0),
+            ROLE_SET_SUBREGISTRY | ROLE_SET_RESOLVER,
+            expires
+        );
+        
+        // Verify registration
+        assertEq(proxy.ownerOf(tokenId), user1, "Domain should be owned by user1");
+        assertEq(proxy.getExpiry(tokenId), expires, "Expiry should match");
+    }
+    
+    // Test for contract upgradeability
+    function test_upgrade() public {
+        // Deploy a new implementation
+        UserRegistryV2Mock newImplementation = new UserRegistryV2Mock();
+        
+        // Upgrade the proxy
+        vm.prank(admin);
+        proxy.upgradeToAndCall(address(newImplementation), "");
+        
+        // Test the new functionality
+        UserRegistryV2Mock upgradedProxy = UserRegistryV2Mock(address(proxy));
+        assertEq(upgradedProxy.version(), 2, "Version should be 2 after upgrade");
+    }
+    
+    function test_Revert_unauthorized_upgrade() public {
+        // Deploy a new implementation
+        UserRegistryV2Mock newImplementation = new UserRegistryV2Mock();
+        
+        // User1 tries to upgrade without permission
+        vm.prank(user1);
+        vm.expectRevert(); // Should revert because user1 doesn't have upgrade rights
```

Comment:
This should test for the exact reversion error so that we know it's the revert we expect.